### PR TITLE
QueryEngine improvements and fixes

### PR DESCRIPTION
**Pull request by Aloren**
- make Index fully immutable
- fix index type to support geo json
- fix index to support nullable set

### CODE REVIEW

**Comment by Aloren** [comment]
@mzhangsfly the changes that I was talking about in the previous PR.
I've also fixed a bunch of issues in index parsing logic and added additional tests.

**Comment by BrianNichols** [comment]
We had to revert the initial merge.  Can you re-issue the full pull request including your latest changes.

**Comment by Aloren** [comment]
@BrianNichols this pr already contains both commits

File: extensions/query-engine/src/main/java/com/aerospike/helper/query/cache/IndexInfoParser.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        }
+        if (indexTypeString.equalsIgnoreCase(STRING_TYPE))
+            return IndexType.STRING;
+        else if (indexTypeString.equalsIgnoreCase(GEO_JSON_TYPE))
+            return IndexType.GEO2DSPHERE;
+        else if (indexTypeString.equalsIgnoreCase(NUMERIC_TYPE))
+            return IndexType.NUMERIC;
+        return null;
+//        TODO: should we throw exception in case of unknown index type?
+//        throw new IllegalStateException("Unknown index type: " + indexTypeString + " in info: " + values);
```
Comment (Aloren) [inline extensions/query-engine/src/main/java/com/aerospike/helper/query/cache/IndexInfoParser.java]: Thought about the case when new index type is added into the server and users still have old client -- I suppose we should not throw exception then. Maybe add warning?

**Comment by BrianNichols** [comment]
Ok.  We will include these changes in tomorrow's release.

**Comment by Aloren** [comment]
thank you @BrianNichols @mzhangsfly for cooperation

**Comment by BrianNichols** [comment]
Java client 4.3.1 has been released with these changes.

https://www.aerospike.com/download/client/java/4.3.1/

**Comment by BrianNichols** [comment]
In addition, spring-data-aerospike 2.0.0.RELEASE was released to maven central and now references Java client 4.3.1.

**Pull request closed without merging**